gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -140,8 +140,24 @@ public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedE
    * @return A {@link ManagedChannel} that can be used for a single RPC call.
    */
   private ManagedChannel getNextChannel() {
-    int currentRequestNum = requestCount.getAndIncrement();
-    int index = Math.abs(currentRequestNum % channels.size());
+    return getChannel(requestCount.getAndIncrement());
+  }
+
+  /**
+   * Returns one of the channels managed by this pool. The pool continues to "own" the channel, and
+   * the caller should not shut it down.
+   *
+   * @param affinity Two calls to this method with the same affinity returns the same channel. The
+   *     reverse is not true: Two calls with different affinities might return the same channel.
+   *     However, the implementation should attempt to spread load evenly.
+   */
+  ManagedChannel getChannel(int affinity) {
+    int index = affinity % channels.size();
+    index = Math.abs(index);
+    // If index is the most negative int, abs(index) is still negative.
```

Comment:
`Math.abs(Integer.MIN_VALUE)` is an interesting case, but if that is possible (theoretically it is, after integer overflow when `requestCount.get() == Integer.MAX_VALUE` and then `requestCount.incrementAndGet()`), then `reuestCount` should be `AtomicLong` instead of `AtomicInteger`. Even if integer overflow is magically handled it would allow a variable called `requestCount` to have negative value, which would make no sense.

In other words, please change requestCount to be AtomicLong (and in general the whole affinity thing should be of type `long`, as in heavy-duty applications with uptime of months/years it should be possible to make more than 2 billion requests in total), then this method can be modified to (comments are for review purpose only):
```java
  private ManagedChannel getChannel(long affinity) {
    // 1) channels.size() is way below Integer.MAX_VALUE the result is guaranteed to fit into int.
    // 2) No need to care about affinity < 0 (overflow): number of seconds since the Big Bang is 10^17, 
    //    maximum long value is higher than 10^18.
    return channels.get((int)(affinity % channels.size()));
  }
```

Reply:
IMHO renaming `requestCount` to something else like `requestTicker` would be more appropriate. The `ChannelPool` doesn't actually care how many requests passed through it, it only needs a circular queue. Using integer overflow for implementing a circular queue is pretty reasonable

Reply:
I went with @igorbernstein2 suggestion. IMHO, properly dealing with negative numbers is a more robust design. An unlikely use-case, but if you want to randomize the channel, you can set affinity to random int and we'll still work.

Reply:
In that case it somewhat complicates the concept of affinity and how it is used. When `requestCounter` is an actual `requestCounter` then it is pretty straigtforward to understand: we have a counter, which keeps total number of calls and affinity is equal to the number of calls made. With request ticker we have two misterious entities: ticker, which is a counter, but not exactly and affinity, which is a blackbox integer value. I also liked that long allows to make this method a oneliner.

It is still a minor thing, and, since you and Igor are in agreement I'm ok with keeping the thing as is.